[MODCON - 65] - Implement endpoint to save shared setting uuid in all tenants

azizbekxm opened the pull request:
<!--
  If you have a relevant JIRA issue number, please put it in the issue title.
  Example: MODCONSORTIA-01 Logging Improvement
  TL;DR
    - https://www.youtube.com/watch?v=5aHmO_S8FQ4
    - http://www.olitreadwell.com/2016/05/22/how-to-write-great-pull-requests/
    - https://www.atlassian.com/blog/git/written-unwritten-guide-pull-requests
-->

## Purpose
<!--
  Why are you making this change? There is nothing more important
  to provide to the reviewer and to future readers than the cause
  that gave rise to this pull request. Be careful to avoid circular
  statements like "the purpose is to update the schema."
  instead, provide an explanation like "there is more data to be provided and stored for Purchase Orders
  which is currently missing in the schema"
  The purpose may seem self-evident to you now, but the standard to
  hold yourself to should be "can a developer parachuting into this
  the project reconstructs the necessary context merely by reading this
  section."
  If you have a relevant JIRA issue, add a link directly to the issue URL here.
  Example: https://issues.folio.org/browse/MODCONSORTIA-01
 -->
https://issues.folio.org/browse/MODCON-65
## Approach
<!--
 How does this change fulfill the purpose? It's best to talk
 high-level strategy and avoid code-splaining the commit history.
 The goal is not only to explain what you did but help other
 developers *work* with your solution in the future.
-->
I will implement 
```
Endpoint contract:
POST /consortia/<consortiumId>/sharing/settings
{
    "settingId": "<UUID>",
    "url": "string",
    "payload": <<object>>
}
Response:
{
    "createSettingsPcId": "<UUID>",
    "updateSettingsPcId": "<UUID>"
}
```
based on this diagram
![image](https://github.com/folio-org/mod-consortia/assets/113523904/e48a2e3a-f737-4e14-a9e5-5ca62ebe3b01)

#### TODOS and Open Questions
<!-- OPTIONAL
- [ ] Use GitHub checklists. When solved, check the box and explain the answer.
- [ ] Check logging.
-->

## Learning
<!-- OPTIONAL
  Help out not only your reviewer but also your fellow developer!
  Sometimes there are key pieces of information that you used to come up
  with your solution. Don't let all that hard work go to waste! A
  pull request is a *perfect opportunity to share the learning that
  you did. Add links to blog posts, patterns, libraries or add-ons used
  to solve this problem.
-->

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [ ] Code coverage on new code is 80% or greater
  - [ ] Duplications on new code is 3% or less
  - [ ] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added or removed?
  - [ ] Are there new interface dependencies?
  - [ ] There are no breaking changes in this PR.

If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all the appropriate links to blocked/related issues?
- Are the JIRAs under active development?
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?
- Did you modify code to call some additional endpoints?
  - [ ] If so, do you check that necessary module permission added in ModuleDescriptor-template.yaml file?

Ideally, all the PRs involved in breaking changes would be merged on the same day
to avoid breaking the folio-testing environment.
Communication is paramount if that is to be achieved,
especially as the number of inter-module and inter-team dependencies increase.

While it's helpful for reviewers to help identify potential problems,
ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

SerhiiNosko commented on src/main/java/org/folio/consortia/controller/SharingSettingController.java:
```diff
@@ -0,0 +1,30 @@
+@RequiredArgsConstructor
+public class SharingSettingController implements SharingApi {
+
+  private final SharingSettingService sharingsettingService;
```
variable should be renamed to sharingSettingService

azizbekxm commented:
```
15:23:11 http-nio-8081-exec-5 [363206/consortia] [diku] [4f53dd76-d8e3-5cf0-a4e2-b51550acb037] [mod-consortia] INFO  ngSettingServiceImpl start:: tenants with size: 2 successfully added to appropriate publication request
15:23:11 http-nio-8081-exec-5 [363206/consortia] [diku] [4f53dd76-d8e3-5cf0-a4e2-b51550acb037] [mod-consortia] INFO  ngSettingServiceImpl start:: set source as a consortium
15:23:11 http-nio-8081-exec-5 [] [diku] [] [mod-consortia] INFO  AuthService          Attempt login with url=http://localhost:9130 tenant=diku username=consortia-system-user.
15:23:11 http-nio-8081-exec-5 [] [diku] [] [mod-consortia] INFO  AuthService          Logged in as consortia-system-user in tenant diku
15:23:22 http-nio-8081-exec-5 [] [diku] [5c780098-751b-4477-9b61-0f75bc452a73] [mod-consortia] INFO  blicationServiceImpl Created publication record bdfd4259-56d1-42ee-9640-6f696d7ee2fc
15:23:24 ConsortiaAsync-2 [] [diku] [5c780098-751b-4477-9b61-0f75bc452a73] [mod-consortia] INFO  TenantContextUtils   FOLIO context initialized with tenant diku
15:23:24 pool-3-thread-1 [] [diku] [5c780098-751b-4477-9b61-0f75bc452a73] [mod-consortia] INFO  TenantContextUtils   FOLIO context initialized with tenant diku
15:23:24 pool-3-thread-1 [] [diku] [5c780098-751b-4477-9b61-0f75bc452a73] [mod-consortia] INFO  TenantContextUtils   FOLIO context initialized with tenant diku
15:23:25 ConsortiaAsync-2 [] [diku] [5c780098-751b-4477-9b61-0f75bc452a73] [mod-consortia] INFO  TenantContextUtils   FOLIO context initialized with tenant diku
15:23:25 ConsortiaAsync-2 [] [diku] [5c780098-751b-4477-9b61-0f75bc452a73] [mod-consortia] INFO  blicationServiceImpl updatePublicationsStatus:: updated publication record bdfd4259-56d1-42ee-9640-6f696d7ee2fc with status ERROR
15:23:25 http-nio-8081-exec-5 [] [diku] [5c780098-751b-4477-9b61-0f75bc452a73] [mod-consortia] INFO  blicationServiceImpl Created publication record d3cf7066-02ee-4926-a37b-7e30ac552554
15:23:28 ConsortiaAsync-3 [] [diku] [5c780098-751b-4477-9b61-0f75bc452a73] [mod-consortia] INFO  TenantContextUtils   FOLIO context initialized with tenant diku
15:23:28 pool-4-thread-1 [] [diku] [5c780098-751b-4477-9b61-0f75bc452a73] [mod-consortia] INFO  TenantContextUtils   FOLIO context initialized with tenant testenant7664803557551864676
15:23:28 pool-4-thread-1 [] [diku] [5c780098-751b-4477-9b61-0f75bc452a73] [mod-consortia] INFO  TenantContextUtils   FOLIO context initialized with tenant diku
15:23:28 ConsortiaAsync-3 [] [diku] [5c780098-751b-4477-9b61-0f75bc452a73] [mod-consortia] INFO  TenantContextUtils   FOLIO context initialized with tenant diku
15:23:28 ConsortiaAsync-3 [] [diku] [5c780098-751b-4477-9b61-0f75bc452a73] [mod-consortia] INFO  blicationServiceImpl updatePublicationsStatus:: updated publication record d3cf7066-02ee-4926-a37b-7e30ac552554 with status ERROR

```

SerhiiNosko commented on src/main/java/org/folio/consortia/controller/SharingSettingController.java:
```diff
@@ -0,0 +1,30 @@
+@RequestMapping("/consortia/{consortiumId}/sharing")
+@Log4j2
+@RequiredArgsConstructor
+public class SharingSettingController implements  SettingsApi{
```
formatting here

SerhiiNosko commented on src/main/java/org/folio/consortia/domain/entity/SharingSettingEntity.java:
```diff
@@ -0,0 +1,41 @@
+@RequiredArgsConstructor
+@Entity
+@Table(name = "sharing_setting")
+public class SharingSettingEntity {
```
Should extend AuditableEntity

azizbekxm replied:
Fixed

SerhiiNosko commented on src/main/java/org/folio/consortia/service/impl/SharingSettingServiceImpl.java:
```diff
@@ -0,0 +1,95 @@
+    TenantCollection tenantCollection = tenantService.getAll(consortiumId);
+
+    PublicationRequest publicationRequestPostMethod = createPublicationRequestForSetting(sharingSettingRequest, "POST");
+    PublicationRequest publicationRequestPutMethod = createPublicationRequestForSetting(sharingSettingRequest, "PUT");
```
there is should be some utitlity class from spring lik HttpStatusCode in the classpath to use constants for http methods

azizbekxm replied:
Used

SerhiiNosko commented on src/main/java/org/folio/consortia/service/impl/SharingSettingServiceImpl.java:
```diff
@@ -0,0 +1,95 @@
+        publicationRequestPostMethod.getTenants().add(tenant.getId());
+      }
+    }
+    log.info("start:: tenants with size: {} successfully added to appropriate publication request", tenantCollection.getTotalRecords());
```
please add log in which tenants you are going to create setting and in which - to update

SerhiiNosko commented on src/main/java/org/folio/consortia/service/impl/SharingSettingServiceImpl.java:
```diff
@@ -0,0 +1,95 @@
+    var updatedPayload = setSourceAsConsortium(payload);
+    publicationRequestPostMethod.setPayload(updatedPayload);
+    publicationRequestPutMethod.setPayload(updatedPayload);
+    log.info("start:: set source as a consortium");
```
log without identifiers is not very informative, you can add identifiers or delete log

SerhiiNosko commented on src/main/java/org/folio/consortia/service/impl/SharingSettingServiceImpl.java:
```diff
@@ -0,0 +1,95 @@
+  }
+
+  private UUID publishRequest(UUID consortiumId, PublicationRequest publicationRequest) {
+    if (!publicationRequest.getTenants().isEmpty()) {
```
CollectionUtils.isNotEmpty

azizbekxm replied:
fixed

SerhiiNosko commented on src/main/java/org/folio/consortia/service/impl/SharingSettingServiceImpl.java:
```diff
@@ -0,0 +1,95 @@
+    log.debug("start:: Trying to share setting with consortiumId: {}, sharing settingId: {}", consortiumId, sharingSettingRequest.getSettingId());
+    consortiumService.checkConsortiumExistsOrThrow(consortiumId);
+    Set<String> tenantsAssociationWithSetting = sharingSettingRepository.findTenantsBySettingId(sharingSettingRequest.getSettingId());
+    TenantCollection tenantCollection = tenantService.getAll(consortiumId);
```
There is some proposal for names:
tenantsAssociationWithSetting -> settingTenants
tenantCollection -> allTenants
publicationRequestPostMethod - publishPostRequest
publicationRequestPutMethod - publishPutRequest

azizbekxm replied:
names was changed

azizbekxm commented on src/test/java/org/folio/consortia/service/SharingSettingServiceTest.java:
```diff
@@ -0,0 +1,94 @@
+
+@SpringBootTest
+class SharingSettingServiceTest {
+  private static final String SHARING_SETTING_REQUEST_SAMPLE = "mockdata/sharing_setting_request.json";
```
added

okolawole-ebsco commented on src/main/java/org/folio/consortia/service/impl/SharingInstanceServiceImpl.java:
```diff
@@ -80,8 +77,7 @@ public SharingInstance start(UUID consortiumId, SharingInstance sharingInstance)
       }
 
       try (var context = new FolioExecutionContextSetter(prepareContextForTenant(targetTenantId, folioModuleMetadata, folioExecutionContext))) {
-        String source = FOLIO_SOURCE_VALUE.equalsIgnoreCase(inventoryInstance.get("source").asText()) ? CONSORTIUM_FOLIO : CONSORTIUM_MARK;
```
I am not sure if it still stands but it should be CONSORTIUM_MARC not CONSORTIUM_MARK.

azizbekxm replied:
fixed

okolawole-ebsco commented on src/main/java/org/folio/consortia/utils/HelperUtils.java:
```diff
@@ -3,8 +3,16 @@
 
+  private static final String FOLIO_SOURCE_VALUE = "folio";
+  private static final String CONSORTIUM_FOLIO = "CONSORTIUM-FOLIO";
+  private static final String CONSORTIUM_MARK = "CONSORTIUM-MARC";
```
Change this to CONSORTIUM_MARC

azizbekxm replied:
Changed

okolawole-ebsco commented on src/main/java/org/folio/consortia/utils/HelperUtils.java:
```diff
@@ -17,4 +25,9 @@ public static String randomString(Integer noOfString) {
   }
+
+  public static ObjectNode setSourceAsConsortium(JsonNode payload) {
+    String source = FOLIO_SOURCE_VALUE.equalsIgnoreCase(payload.get("source").asText()) ? CONSORTIUM_FOLIO: CONSORTIUM_MARK;
```
I think we should be more explicit here to map folio -> CONSORTIUM-FOLIO and marc -> CONSORTIUM-MARC. if other source values are introduced in the future, it would default to CONSORTIUM-MARC and that wont be ideal. Probably should return an error if we find a source value that is not recognized.

azizbekxm replied:
1. Do you mean this structure

- if source is folio, then it will be CONSORTIUM-FOLIO
- If source is marc, then it will be CONSORTIUM-MARC.
- all other type will throw exceptino
![image](https://github.com/folio-org/mod-consortia/assets/113523904/a87b328f-4c22-461a-a9e3-27edc3851768)

2. What about instance? is it still valid scenario for sharing instance (this code is used by share instance method also)

okolawole-ebsco replied:
1. The screenshot looks incorrect to me because it is saying if source = CONSORTIUM_FOLIO, set source to CONSORTIUM_FOLIO. But the words are what I expect.

2. I was not aware that this was the case. it was my impression that we would save "consortium" instead of "consortium-folio"  for settings. I was thinking about my initial comment in the instance context instead of settings. 

Generally, what I am looking for is that CONSORTIUM-MARC is not the default output when source does not match CONSORTIUM-FOLIO.

SerhiiNosko replied:
Yes agree, @azizbekxm for instances if source = "folio" - result should be "CONSORTIUM_FOLIO", if source = "marc" - result should be "CONSORTIUM_MARC"
For Settings it always should be "consortium"

SerhiiNosko replied:
@okolawole-ebsco maybe we can use lower case as "consortium_folio" or "consortium_marc" to make it consistent?

SerhiiNosko replied:
@azizbekxm handling such situation using flag isSetting is not approprtiate here, you can leave constants in utility class and move business logic to appropriate service

azizbekxm replied:
Fixed big thanks

SerhiiNosko commented on src/main/java/org/folio/consortia/service/impl/SharingInstanceServiceImpl.java:
```diff
@@ -80,7 +79,11 @@ public SharingInstance start(UUID consortiumId, SharingInstance sharingInstance)
       try (var context = new FolioExecutionContextSetter(prepareContextForTenant(targetTenantId, folioModuleMetadata, folioExecutionContext))) {
-        String source = FOLIO_SOURCE_VALUE.equalsIgnoreCase(inventoryInstance.get("source").asText()) ? CONSORTIUM_FOLIO : CONSORTIUM_MARK;
+        String source = switch (inventoryInstance.get("source").asText()) {
+          case "folio" -> CONSORTIUM_FOLIO;
```
theretically souce can come with upper registry, so safer to compare with equalsIgnoreCase

azizbekxm replied:
Fixed by using .toLowerCase method of string

SerhiiNosko commented on src/main/java/org/folio/consortia/service/impl/SharingSettingServiceImpl.java:
```diff
@@ -0,0 +1,118 @@
+    for (Tenant tenant : allTenants.getTenants()) {
+      if (settingTenants.contains(tenant.getId())) {
+        publicationPutRequest.getTenants().add(tenant.getId());
+        log.info("start:: tenant={} added to publication update request for setting={}", tenant.getId(), settingId);
```
you also need to update existing setting ids in database to have correct metadata when they updated

SerhiiNosko commented on src/main/java/org/folio/consortia/service/impl/SharingSettingServiceImpl.java:
```diff
@@ -0,0 +1,118 @@
+      } else {
+        publicationPostRequest.getTenants().add(tenant.getId());
+        log.info("start:: tenant={} added to publication create request for setting={}", tenant.getId(), settingId);
+        saveSharingSettingEntity(sharingSettingRequest, tenant.getId());
```
Accumulate all entities in list and invoke saveAll() in spring data

azizbekxm replied:
Fixed

tatsiana-tarhonskaya commented on src/main/java/org/folio/consortia/service/impl/SharingSettingServiceImpl.java:
```diff
@@ -0,0 +1,123 @@
+    log.info("start:: The Sharing Settings for settingId '{}' and '{}' unique tenant(s) were successfully saved to the database",
+      sharingSettingRequest.getSettingId(), publicationPostRequest.getTenants().size());
+
+    ObjectMapper objectMapper = new ObjectMapper();
```
this should be a class field

azizbekxm replied:
Moved to class field
